google-ads-examples/src/main/java/com/google/ads/googleads/examples/shoppingads/GetProductBiddingCategoryConstant.java

```diff
@@ -0,0 +1,193 @@
+      List<CategoryNode> rootCategories = new ArrayList<>();
+      // Creates a map of category ID to category node for all categories found in the results.
+      HashMap<String, CategoryNode> biddingCategories = new HashMap<>();
+
+      // Performs the search request.
+      SearchPagedResponse response = googleAdsServiceClient.search(request);
+      for (GoogleAdsRow googleAdsRow : response.iterateAll()) {
+        ProductBiddingCategoryConstant productBiddingCategory =
+            googleAdsRow.getProductBiddingCategoryConstant();
+
+        String name = productBiddingCategory.getLocalizedName().getValue();
+        String resourceName = productBiddingCategory.getResourceName();
+        CategoryNode node = biddingCategories.get(resourceName);
+        if (node == null) {
+          node = new CategoryNode(resourceName, name);
+          biddingCategories.put(resourceName, node);
+        } else if (node.name == null) {
+          // Ensures that the name attribute for the node is set. Name will be null for nodes added
+          // to biddingCategories as a result of being a parentNode below.
+          node.name = name;
+        }
+
+        if (productBiddingCategory.hasProductBiddingCategoryConstantParent()) {
+          String parentResourceName =
+              productBiddingCategory.getProductBiddingCategoryConstantParent().getValue();
+          CategoryNode parentNode = biddingCategories.get(parentResourceName);
+          if (parentNode == null) {
+            parentNode = new CategoryNode(parentResourceName);
+            biddingCategories.put(parentResourceName, parentNode);
+          }
+          parentNode.children.add(node);
+        } else {
+          rootCategories.add(node);
+        }
+      }
+      displayCategories(rootCategories, "");
+    }
+  }
+
+  /**
+   * Recursively prints out each category node and its children.
+   *
+   * @param categories the categories to print.
+   * @param prefix the string to print at the beginning of each line of output.
+   */
+  private static void displayCategories(List<CategoryNode> categories, String prefix) {
+    for (CategoryNode category : categories) {
+      System.out.printf("%s%s [%s]%n", prefix, category.name, category.resourceName);
+      displayCategories(category.children, String.format("%s%s > ", prefix, category.name));
+    }
+  }
+
+  /** Node that tracks a product bidding category's id, name, and child nodes. */
+  private static class CategoryNode {
+    final String resourceName;
+    String name;
+    final List<CategoryNode> children;
+
+    /**
+     * Constructor for categories first encountered as non-parent elements in the results.
+     *
+     * @param resourceName the resource name = of the category
+     * @param name the name of the category
+     */
+    CategoryNode(String resourceName, String name) {
+      this.children = new ArrayList<>();
+      this.resourceName = Preconditions.checkNotNull(resourceName);
```

Comment:
nit: use the [version of checkNotNull](https://google.github.io/guava/releases/snapshot/api/docs/com/google/common/base/Preconditions.html#checkNotNull-T-java.lang.Object-) that also takes a string that explains what's unexpectedly null. Otherwise, the resulting NPE won't have a message.